qubes/vm/dispvm.py

```diff
@@ -187,11 +189,130 @@ def __init__(self, app, xml, *args, **kwargs):
             self.features.update(template.features)
             self.tags.update(template.tags)
 
+    def is_domain_preloaded(self):
+        preload_dispvm = (
+            self.features.check_with_template("preload-dispvm", None)
+        )
+        if not preload_dispvm:
+            return False
+        preload_dispvm = preload_dispvm.split(" ")
+        if self.name not in preload_dispvm:
+            return False
+        return True
+
+    async def create_preloaded_dispvm(self):
+        """
+        Create preloaded DispVM.
+
+        Template from which the VM should be created.
+
+        :return:
+        """
+        preload_dispvm_max = int(
+            self.features.check_with_template("preload-dispvm-max", 0)
+        )
+        if preload_dispvm_max == 0:
+            return
+
+        preload_dispvm = self.features.check_with_template(
+            "preload-dispvm", None
+        )
+        if preload_dispvm:
+            if len(preload_dispvm.split(" ")) >= preload_dispvm_max:
+                return
+            preload_dispvm += " " + self.name
+        else:
+            preload_dispvm = self.name
+
+        dispvm_template = getattr(self, "template")
+        dispvm_template.features["preload-dispvm"] = preload_dispvm
+        self.features["internal"] = True
+
+    async def use_preloaded_dispvm(self):
+        """
+        Mark preloaded DispVM as used.
+
+        Template from which the VM should be used (first preloaded VM) or the
+        VM itself.
+
+        :return:
+        """
+        if self.klass == "DispVM":
+            dispvm_template = getattr(self, "template")
+            dispvm = self
+        elif (
+            self.klass == "AppVM"
+            and getattr(self, "template_for_dispvms", False)
+        ):
+            dispvm_template = dispvm
+        else:
+            raise qubes.exc.QubesException("Invalid argument provided")
+
+        preload_dispvm = dispvm_template.features.get("preload-dispvm", None)
+        if not preload_dispvm:
+            return
+
+        preload_dispvm = preload_dispvm.split(" ")
+        if self.klass == "DispVM":
+            if dispvm.name not in preload_dispvm:
+                raise qubes.exc.QubesException("DispVM is not preloaded")
+            used_dispvm = dispvm.name
+        else:
+            used_dispvm = preload_dispvm[0]
+            dispvm = self.app.domains[used_dispvm]
+
+        unused_dispvm = preload_dispvm.remove(used_dispvm)
+        dispvm_template.features["preload-dispvm"] = unused_dispvm
+        dispvm.features["internal"] = False
+        ## TODO: confirm if this guess of event code is correct
+        self.app.fire_event_for_permission(dispvm_template=dispvm_template)
+        dispvm_template.fire_event_async("domain-preloaded-dispvm-used")
```

Comment:
_async needs `await`. And also, it may be useful to add dispvm (name? object?) as an event parameter.
And please document the event (there is specific sphinx syntax for events, see others for example).